File: PA94.java
//Leetcode: 378. Kth Smallest Element in a Sorted Matrix
//Time complexity: O(nlogk) // where n is the number of elements in the matrix and k is the size of heap
//Space Complexity: O(k) // k is the size of the heap
class Solution {
    public int kthSmallest(int[][] matrix, int k) {
        if(matrix==null   || matrix.length==0 || matrix[0].length==0) return -1;
        
        PriorityQueue<Integer> pq= new PriorityQueue<>((a,b)->(b-a));
        
        for(int i=0;i<matrix.length;i++){
            for(int j=0;j<matrix[0].length;j++){
                pq.offer(matrix[i][j]);
                if(pq.size()>k){
                    pq.poll();
                }
            }
        }       
        return pq.poll();
        
        
    }
}

File: PA94.py
#Time Complexity - O(m*n)
#Space Complexity - O(k)
import heapq
class Solution(object):
    def kthSmallest(self, matrix, k):
        """
        :type matrix: List[List[int]]
        :type k: int
        :rtype: int
        """
        q = []
        for i in range(len(matrix)):
            for j in range(len(matrix)):
                heapq.heappush(q,matrix[i][j]*-1)
                if len(q) > k:
                    heapq.heappop(q)
        return heapq.heappop(q)*-1